### ISSUE TITLE
gas errors when using bridge with private geth network

### ISSUE THREAD

**Issue by IamChuancey** [oraclize/ethereum-bridge#72]
I am using the ethereum-bridge to connect oraclize with my local private network,but it seems that it didn't work at all!!!
![image](https://user-images.githubusercontent.com/29730996/56865639-ec648400-6a02-11e9-95f2-2d73605bc891.png)
So,i am successfully building  ethereum-bridge,this is my contract deployed on Remix
![image](https://user-images.githubusercontent.com/29730996/56865678-377e9700-6a03-11e9-9ef5-b3acaa8a8584.png)
i can see that the oraclize_query never sends out at all!

**Comment by IamChuancey** [comment]
I have achieved some progress on this issue, it seems that it works well on testnets like Ropsten or Kovan, but it still can not work on local private network that using geth.

**Comment by D-Nice** [comment]
We'll need some more information on how you're running this. The bridge is used for a number of projects and does work. The bridge generally sees most of its use with truffle + ganache(testrpc). 

For quick prototyping, note that you can use the Oraclize plugin, available directly on remix.ethereum.org with the javascript VM set, so you don't need to run your own geth instance, which I'm guessing is the misconfigured part.

Once wanting to move to a local/private net workflow, I recommend basing it off our https://github.com/oraclize/truffle-starter repo.

**Comment by IamChuancey** [comment]
Thank you very much! Yeah, I have succeed in connecting Oraclize in javascript VM set with the  the Oraclize plugin on.And also, i have succeed on testnets like Ropsten or Kovan, but it remains unworking in local/private network using geth like `geth --networkid 17 --datadir ./data --rpc --rpcapi "admin,debug,eth,miner,net,personal,txpool,web3" --rpc --rpcport "8545" --rpccorsdomain "*" `(this is the way to start my private network).I am confused why private network is not working as the same OAR is working in testnets but not in my private network. BTW, I use Remix to deploy my contract.

**Comment by IamChuancey** [comment]
And i think it's something about the gas in private network using geth,beacuse it always the same error:`gas required exceeds allowance or always failing transaction in private network`,so i set the gas limit and the eth value,but nothing changed! i need your help.

**Comment by gskapka** [comment]
Hi @IamChuancey - I've written much the same in our support channel on gitter but will post it here also incase you aren't monitoring that channel as much.

At a glance, there are two potential culprits: 

- The first could be your genesis json file, and what the block gas limit is set to in there. The live public test nets have higher block gas limits now then when they were begun, and your contracts along with the Provable API might be deployable on the test-nets but not your private network due to this difference in the block gas limit.

- The second potential issue is where you mention: 

> I am confused why private network is not working as the same OAR is working in testnets but not in my private network.

By this do you mean that the OAR you are attempting to use with your private Geth network is the same OAR that you've used and found to be working on the public test nets? 

Because if so, that wouldn't work since the OAR addresses on the public test-nets are ones that _we_ have deployed ourselves to those networks, and so they won't exist on your private network. 

If that is indeed the case, you'll have to start the `ethereum-bridge` and let it deploy an OAR and then you can use _that_ address in your private network contracts. The output of the bridge will echo this OAR address once it's deployed and the bridge is up and running, and give you the string you need to copy and paste into your contract's constructor(s).

**Comment by IamChuancey** [comment]
I am so appreciate that you can help, but it seems it is not the mistake that you mentioned above. I have checked that  the block gas limit is high enough and I have changed the OAR but nothing changed.
I can concur that  it is something wrong about oraclize_query,beacuse the transaction when I call the query function didn't emit the event I defined in the query function.
`function queryPriceofLPG() public payable{
       emit LogNewOraclizeQuery("Oraclize query was sent, standing by for the answer..");
       oraclize_query("WolframAlpha","random number between 1 and 10");
}`
![image](https://user-images.githubusercontent.com/29730996/56938244-65fb8f80-6b33-11e9-9ced-d4a6f6dda418.png)

**Comment by IamChuancey** [comment]
by the way ,the gas limit in my genesis file is set to 0xffffffff, so it will not be the problem about gas limit

**Comment by IamChuancey** [comment]
I can show you more detail:
This is my code:
```
pragma solidity ^0.5.1;
import "github.com/oraclize/ethereum-api/oraclizeAPI.sol";
contract ExampleContract is usingOraclize{
string public EURGBP;
event updatedPrice(string price);
event newOraclizeQuery(string description);

constructor()public  {
    OAR = OraclizeAddrResolverI(0x8F329769ef209af0E6b4f8f92CB64De6C1912182);
}

function __callback(bytes32 myid, string memory result) public{
    require(msg.sender==oraclize_cbAddress(),"you are not allowed!");
    EURGBP = result;
   emit updatedPrice(result);
}

function updatePrice() payable public{
    if (oraclize_getPrice("URL") > msg.value) {
        emit newOraclizeQuery("Oraclize query was NOT sent, please add some ETH to cover for the query fee");
    } else {
      emit  newOraclizeQuery("Oraclize query was sent, standing by for the answer..");
      oraclize_query("WolframAlpha","the temperature in London");
    }
}
}
```
and the result is every time I call the function`updatePrice`,it will always show me:
![image](https://user-images.githubusercontent.com/29730996/56946949-44fc6400-6b5e-11e9-9e1a-930dc339294c.png)
 and when I click on send transaction, the transaction is sending out, but it didn't emit the events defined in the function which means the function didn't work at all.

**Comment by gskapka** [comment]
Could you show the console output from when you start the `ethereum-bridge`, as well as the string you start the bridge with?

**Comment by IamChuancey** [comment]
first of all, really thank you.
this is the console output from when i start the ethereum-bridge:
![image](https://user-images.githubusercontent.com/29730996/56965180-4003d880-6b8f-11e9-889f-2faa3cd07df1.png)
so when in this situation, should i add the OAR statement in my contract?
` OAR = OraclizeAddrResolverI(0x6f485c8bf6fc43ea212e93bbf8ce046c7f1cb475);`
if it's needed,which address should i use? the connector address`0x70cc4b3a8af0a4c3e7451529b21497ee4c7b490e` or the oar address `0x6f485c8bf6fc43ea212e93bbf8ce046c7f1cb475`?, but when i use either address of the two, it all didn't work.

**Comment by gskapka** [comment]
You should definitely use the OAR address, so `0x6f...` in this case. The OAR is simply a proxy to the actual connector contract which does all the heavy lifting. 

When you deploy a contract that's using the Provable API, it'll look for the OAR at any addresses it has hardcoded in it (so for the main net and public testnets), and when you add the OAR to your contracts constructor it'll also check that one. 

Using that method, it'll find the connector contract and thus be able to run functions in that contract, by way of your API using contract.

So, with that in mind, there appears to be some error when you call your function which calls an `oraclize_query` method, and that error is manifasting as the gas-estimation-error. This can often mean a `revert()` is being hit, hence the always failing part. 

As for what precisely is failing, that's a little more difficult to pin down. So first we need to check all the low handing fruit: 

- Have you tried sending ETH with the request?
- Have you tried setting the gas limit of the transaction to something huge like `6,000,000`?
- Do you see anything in the `ethereum-bridge`'s console when you run the `updatePrice` function in your contract?

**Comment by IamChuancey** [comment]
-  i have sending enough ETH with the request
-  i have tried setting the gas limit of the transaction to something huge like `4294967295`
- the` ethereum-bridge`'s console didn't change when i run the` updatePrice` function:
![image](https://user-images.githubusercontent.com/29730996/56971944-b9ee8e80-6b9c-11e9-837f-4af4345c257f.png)
- and the `bridge.log` is 
![image](https://user-images.githubusercontent.com/29730996/56972150-0c2faf80-6b9d-11e9-8b0c-988892ead09e.png)

**Comment by gskapka** [comment]
Are you loading the bridge from an instance file or from fresh each time? Same question for your geth network, is it a brand new blockchain each time?

**Comment by D-Nice** [comment]
please run the bridge with

`ethereum-bridge -H localhost:8545 --broadcast -a 1`

assuming your node is running locally on port 8545 and the callback account you wish to use is account 1. You must use a different account to deploy your Oraclize contracts. Broadcast mode is required for private chains.

**Comment by IamChuancey** [comment]
yeah, thank you for your effort! And I finally take the first step.
The gas error is something about Internal mechanism of the Remix that causes, it's not the fault of the program or the bridge. 
And when I move my eyes to `web3.js` to interact with my contract , there is no more mistakes about the gas, but it is still wired that I can't get any result return back, the result remains null even if I know it need some time to wait as the oraclize is asynchronous.

**Comment by IamChuancey** [comment]
if possible, you can repeat the procedure as referred in #73

**Comment by IamChuancey** [comment]
what is the status when the bridge works normally?will there be any queries passed in the ethereum bridge on the ethereum-bridge console? because whatever i done, there is no change on the bridge console and bridge.log. Can i understand the situation is the query didn't send out and passed in the bridge?

**Comment by gskapka** [comment]
When the bridge is working normally you will see the query appear in the console as soon as you make a query via your smart-contract (since the bridge has seen the event fired by the connector-contract). 

You wil then see the bridge enter a loop checking for when your query has been exectured by Provable, at which point it will create and send the `__callback` transaction. You'll see this happen too from the bridge. If the `__callback` errors for some reason, you see that happen as well, along with the error message.

So yes you are understanding correctly - if you aren't seeing your queries appear in the bridge as soon as you make them via your smart-contract then there is something not correct about your set-up.

**Comment by D-Nice** [comment]
The bridge output should look like this (bottom tmux session) when it's working successfully.

![2019-05-01-124728_2018x1021_scrot](https://user-images.githubusercontent.com/2888248/57029413-af211180-6c0f-11e9-82c6-c2ec906bc418.png)

Closing this issue as it's a duplicate of https://github.com/oraclize/ethereum-bridge/issues/73 and has an answer there that the bridge supports private chains, confirmed by us, but when it comes to the configuration, if you use a non-default setup, it's up to you to have it setup correctly. One missing thing for example is byzantiumFork which may affect things.

**Comment by IamChuancey** [comment]
The problem is solved in #73
